google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc_test.cc

```diff
@@ -0,0 +1,462 @@
+}
+
+class ResumableAsyncReadWriteStreamingRpcTest : public ::testing::Test {
+ protected:
+  ResumableAsyncReadWriteStreamingRpcTest()
+      : backoff_policy_{std::make_shared<StrictMock<MockBackoffPolicy>>()},
+        stream_{MakeResumableAsyncStreamingReadWriteRpcImpl<FakeRequest,
+                                                            FakeResponse>(
+            retry_policy_factory_.AsStdFunction(), backoff_policy_,
+
+            sleeper_.AsStdFunction(), stream_factory_.AsStdFunction(),
+            initializer_.AsStdFunction())} {}
+
+  StrictMock<MockFunction<future<void>(std::chrono::milliseconds)>> sleeper_;
+  StrictMock<MockFunction<AsyncReadWriteStreamReturnType()>> stream_factory_;
+  StrictMock<MockFunction<std::unique_ptr<MockRetryPolicy>()>>
+      retry_policy_factory_;
+  StrictMock<MockFunction<future<StatusOr<AsyncReadWriteStreamReturnType>>(
+      AsyncReadWriteStreamReturnType)>>
+      initializer_;
+  std::shared_ptr<StrictMock<MockBackoffPolicy>> backoff_policy_;
+  ResumableAsyncReadWriteStream stream_;
+};
+
+TEST_F(ResumableAsyncReadWriteStreamingRpcTest, SingleStartFailureThenGood) {
+  InSequence seq;
+
+  auto stream1 = absl::make_unique<StrictMock<AsyncReaderWriter>>();
+  auto& stream1_ref = *stream1;
+
+  auto mock_retry_policy = absl::make_unique<StrictMock<MockRetryPolicy>>();
+  auto& retry_policy_ref = *mock_retry_policy;
+  EXPECT_CALL(retry_policy_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(mock_retry_policy))));
+
+  auto backoff_policy = absl::make_unique<StrictMock<MockBackoffPolicy>>();
+  auto& backoff_policy_ref = *backoff_policy;
+  EXPECT_CALL(*backoff_policy_, clone)
+      .WillOnce(Return(ByMove(std::move(backoff_policy))));
+
+  EXPECT_CALL(stream_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(stream1))));
+
+  promise<bool> start_promise;
+  EXPECT_CALL(stream1_ref, Start)
+      .WillOnce(Return(ByMove(start_promise.get_future())));
+
+  auto status_future = stream_->Start();
+
+  EXPECT_CALL(stream1_ref, Finish)
+      .WillOnce(Return(ByMove(make_ready_future(kFailStatus))));
+  EXPECT_CALL(retry_policy_ref, IsExhausted).WillOnce(Return(false));
+  EXPECT_CALL(retry_policy_ref, OnFailure(kFailStatus)).WillOnce(Return(true));
+  EXPECT_CALL(backoff_policy_ref, OnCompletion)
+      .WillOnce(Return(std::chrono::milliseconds(7)));
+  EXPECT_CALL(sleeper_, Call).WillOnce(ZeroSleep());
+
+  auto stream2 = absl::make_unique<StrictMock<AsyncReaderWriter>>();
+  auto& stream2_ref = *stream2;
+  EXPECT_CALL(stream_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(stream2))));
+  EXPECT_CALL(stream2_ref, Start)
+      .WillOnce(Return(ByMove(make_ready_future(true))));
+
+  EXPECT_CALL(initializer_, Call).WillOnce(InitializeInline());
+
+  start_promise.set_value(false);
+
+  // start future doesn't finish until
+  // permanent error or `Finish`is called
+  EXPECT_EQ(status_future.wait_for(ms(kFutureWaitMs)),
```

Comment:
Here is where `kFutureWaitMs` is used.  That blocks the current thread (the testing thread) for at least `kFutureWaitMs`.  If there is no reason to block it for that long, we should make the time shorter, so the test can finish faster.